### ISSUE TITLE
1.13.4 Babylon does not correctly parse TypeScript code

### ISSUE THREAD

**Issue by uxuf** [prettier/prettier#4633]
**Prettier 1.13.4**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEATOYA2BDATnAAgHNMIAjbTA4AHSgIYIEtY5cAzbMQgdRdQgB3anUZiCAQQDCUpAWxQAngG5RjAL511IADQgIABxhNoAZ2Sg8uIQAU8CcymwA3CE1S6QZXFwDWcGABlAy4WImQYXABXOD0WUzYYGx8iAFtsZE5MBL0AK1MADwAhHzB-IOxUuAAZFjhMyhyQENwE3GQvbDJFUihPA1wWGD5UGAALZAAOAAY9AYgEnh8DDoG4Nud6vXwARyimfGTsNIykLKaE1KYI6NiQUzDMOABFKIh4Buy7mC6R8eQAEx6SLYJiYMJSCCpdIdKDQLYgKIJAAqXUc5zg6nUQA)
```sh
--parser babylon
```

**Input:**
```ts
declare global {
    interface Window {
        ACC: any;
    }
}
```

**Output:**
```ts
SyntaxError: Unexpected token (1:9)
> 1 | declare global {
    |         ^
  2 |     interface Window {
  3 |         ACC: any;
  4 |     }
```

**Expected behavior:**
```ts
declare global {
    interface Window {
        ACC: any;
    }
}
```

This works for 1.12.0 and below, and works for 1.13.4 with `--parser typescript`. However I have not seen anything in the changelog that mentions Babylon dropping support for TypeScript.

**Comment by ikatyang** [comment]
IIRC, TypeScript never worked with babylon before since we didn't enable the typescript plugin for it.

I tried `prettier@1.12.0` and it seems it did never work before.

```sh
mkdir prettier-test
cd prettier-test
yarn init -y
yarn add prettier@1.12.0
echo 'declare global { interface Window { ACC: any; }}' | yarn prettier --parser babylon
```

```
[error] stdin: SyntaxError: Unexpected token (1:9)
[error] > 1 | declare global { interface Window { ACC: any; }}
[error]     |         ^
[error]   2 |
```

**Comment by ikatyang** [comment]
I'm not sure what should I do here since your example seems not reproducible to me.

**Comment by uxuf** [comment]
You're obviously right. I wonder if my .ts files get parsed by the JavaScript parser pre 1.13.0 if `--parser` is not specified. Let me try that out.

**Comment by uxuf** [comment]
Ok yes so my *.ts files did get parsed as JavaScript pre 1.13.0 when `--parser` was not specified but a glob was.

```
$  prettier-test echo 'declare global { interface Window { ACC: any; }}' >> test.ts
$  prettier-test ./node_modules/prettier/bin-prettier.js test.ts
declare global {
  interface Window {
    ACC: any;
  }
}```

**Comment by lydell** [comment]
@uxuf Your latest example works with both 1.12 and 1.13.

**Comment by uxuf** [comment]
Is there a way for Prettier to spit out what parser it is using? Is there a `--debug` option?

**Comment by lydell** [comment]
There’s `--loglevel debug` although it doesn’t seem to show anything interesting in this case:

```
$ ./node_modules/.bin/prettier test.ts --loglevel debug
[debug] normalized argv: {"color":true,"editorconfig":true,"loglevel":"debug","plugin-search-dir":[],"plugin":[],"ignore-path":".prettierignore","config-precedence":"cli-override","_":["test.ts"]}
[debug] resolve config from 'test.ts'
[debug] loaded options `null`
[debug] applied config-precedence (cli-override): {}
declare global {
  interface Window {
    ACC: any;
  }
}
```

**Comment by uxuf** [comment]
Okay I did not think it was relevant earlier, but apparently it is. I've got a `.prettierrc`:
```
# .prettierrc
parser: babylon
```
With this, on 1.12.0:
```
➜  prettier-test ./node_modules/prettier/bin-prettier.js test.ts
declare global {
  interface Window {
    ACC: any;
  }
}
```
and on 1.13.0:
```
➜  prettier-test ./node_modules/prettier/bin-prettier.js test.ts
[error] test.ts: SyntaxError: Unexpected token (1:9)
[error] > 1 | declare global { interface Window { ACC: any; }}
[error]     |         ^
[error]   2 |
```

I have a hybrid project with a mixture of JavaScript and TypeScript code. I am not quite sure what to think if the typescript plugin was never enabled for babylon. There is definitely a configuration override  (based on file extension?) that happened pre-1.13.0 which doesn't happen anymore in 1.13.0. As far as I know `declare` is not valid JavaScript.

**Comment by lydell** [comment]
Very interesting.

```
$ echo "parser: babylon" > .prettierrc

$ npm i prettier@1.12
+ prettier@1.12.1
updated 1 package in 0.552s

$ ./node_modules/.bin/prettier test.ts --loglevel debug
[debug] normalized argv: {"color":true,"editorconfig":true,"loglevel":"debug","plugin":[],"ignore-path":".prettierignore","config-precedence":"cli-override","_":["test.ts"]}
[debug] resolve config from 'test.ts'
[debug] loaded options `{"parser":"babylon"}`
[debug] applied config-precedence (cli-override): {"parser":"babylon"}
declare global {
  interface Window {
    ACC: any;
  }
}

$ npm i prettier@1.13
+ prettier@1.13.4
updated 1 package in 0.444s

$ ./node_modules/.bin/prettier test.ts --loglevel debug
[debug] normalized argv: {"color":true,"editorconfig":true,"loglevel":"debug","plugin-search-dir":[],"plugin":[],"ignore-path":".prettierignore","config-precedence":"cli-override","_":["test.ts"]}
[debug] resolve config from 'test.ts'
[debug] loaded options `{"parser":"babylon"}`
[debug] applied config-precedence (cli-override): {"parser":"babylon"}
[error] test.ts: SyntaxError: Unexpected token (1:9)
[error] > 1 | declare global { interface Window { ACC: any; }}
[error]     |         ^
[error]   2 | 
```

Not sure if 1.13 is a regression or a bugfix.

Btw, putting `parser:` at toplevel in .prettierrc is really bad. It basically forces you to only be able to format one language. This should be better:

```
overrides:
- files: "*.js"
  options:
    parser: babylon
```

But babylon is the default anyway, so one could just as well remove it.

So try removing `parser: babylon` from your .prettierrc and things should work.

I wonder if we should forbid `parser: babylon` at the toplevel in .prettierrc since you are not the first one to make that mistake.

**Comment by uxuf** [comment]
I think what has happened here that we started with prettifying JavaScript code. When we started introducing TypeScript the configuration worked well. We have never given it any serious thought that the babylon parser does not parse TypeScript on its own, because it seemed like _it did_.

**What we thought was a feature, was in fact a bug.**

I believe the change should be classified as breaking, as 1.13.0 is not backwards compatible in cases like mine since I'm clearly "not the first one to make that mistake".

In the meanwhile I'll update my `.prettierrc` breaking it down by `*.js` and `*.ts` files and the appropriate parsers.
